File: macos-app/Moku/Formats/HtmlParser.swift
import Foundation

/// Parses HTML files into chapters for WebView rendering.
enum HtmlParser {

    struct HtmlMetadata {
        let title: String
        let author: String
        let chapterCount: Int
    }

    struct HtmlChapter {
        let index: Int
        let title: String
        let content: String // raw HTML body content
    }

    // MARK: - Cache

    private nonisolated(unsafe) static var cache: [String: [HtmlChapter]] = [:]

    static func clearCache(_ key: String) {
        cache.removeValue(forKey: key)
    }

    // MARK: - Metadata

    static func extractMetadata(from url: URL) -> HtmlMetadata {
        guard let data = try? Data(contentsOf: url),
              let text = String(data: data, encoding: .utf8) else {
            return HtmlMetadata(
                title: url.deletingPathExtension().lastPathComponent,
                author: "Unknown",
                chapterCount: 1
            )
        }

        let title = extractTag(from: text, tag: "title")
            ?? url.deletingPathExtension().lastPathComponent
        let author = extractMetaContent(from: text, name: "author") ?? "Unknown"
        let chapters = splitByHeadings(text)
        return HtmlMetadata(title: title, author: author, chapterCount: chapters.count)
    }

    // MARK: - Chapters

    static func getChapters(filePath: String, booksDir: URL) throws -> [ChapterInfo] {
        let chapters = try loadOrCache(filePath: filePath, booksDir: booksDir)
        return chapters.map { ChapterInfo(index: $0.index, title: $0.title) }
    }

    static func getChapterContent(filePath: String, chapterIndex: Int, booksDir: URL) throws -> String {
        let chapters = try loadOrCache(filePath: filePath, booksDir: booksDir)
        guard chapterIndex >= 0, chapterIndex < chapters.count else { return "" }
        return chapters[chapterIndex].content
    }

    // MARK: - Private

    private static func loadOrCache(filePath: String, booksDir: URL) throws -> [HtmlChapter] {
        if let cached = cache[filePath] { return cached }
        let url = booksDir.appendingPathComponent(filePath)
        let data = try Data(contentsOf: url)
        guard let text = String(data: data, encoding: .utf8) else {
            return [HtmlChapter(index: 0, title: "Document", content: String(decoding: data, as: UTF8.self))]
        }
        let chapters = splitByHeadings(text)
        cache[filePath] = chapters
        return chapters
    }

    private static func splitByHeadings(_ html: String) -> [HtmlChapter] {
        // Extract body content if present
        let body: String
        if let bodyRange = html.range(of: "<body", options: .caseInsensitive),
           let closeTag = html[bodyRange.upperBound...].range(of: ">"),
           let endBody = html.range(of: "</body>", options: .caseInsensitive) {
            body = String(html[closeTag.upperBound..<endBody.lowerBound])
        } else {
            body = html
        }

        // Split by <h1> or <h2> tags
        let pattern = try! NSRegularExpression(
            pattern: #"<h[12][^>]*>(.*?)</h[12]>"#,
            options: [.caseInsensitive, .dotMatchesLineSeparators]
        )
        let matches = pattern.matches(in: body, range: NSRange(body.startIndex..., in: body))

        guard !matches.isEmpty else {
            return [HtmlChapter(index: 0, title: "Document", content: body)]
        }

        var chapters: [HtmlChapter] = []
        for (i, match) in matches.enumerated() {
            let titleRange = Range(match.range(at: 1), in: body)!
            let title = String(body[titleRange])
                .replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
                .trimmingCharacters(in: .whitespacesAndNewlines)

            let start = Range(match.range, in: body)!.lowerBound
            let end: String.Index
            if i + 1 < matches.count {
                end = Range(matches[i + 1].range, in: body)!.lowerBound
            } else {
                end = body.endIndex
            }
            let content = String(body[start..<end])
            chapters.append(HtmlChapter(index: i, title: title.isEmpty ? "Section \(i+1)" : title, content: content))
        }

        // If there's content before the first heading, prepend it
        let firstMatchStart = Range(matches[0].range, in: body)!.lowerBound
        let prefix = body[body.startIndex..<firstMatchStart].trimmingCharacters(in: .whitespacesAndNewlines)
        if !prefix.isEmpty {
            chapters.insert(HtmlChapter(index: 0, title: "Introduction", content: prefix), at: 0)
            // Re-index
            chapters = chapters.enumerated().map { (i, ch) in
                HtmlChapter(index: i, title: ch.title, content: ch.content)
            }
        }

        return chapters.isEmpty ? [HtmlChapter(index: 0, title: "Document", content: body)] : chapters
    }

    private static func extractTag(from html: String, tag: String) -> String? {
        let pattern = try! NSRegularExpression(
            pattern: "<\(tag)[^>]*>(.*?)</\(tag)>",
            options: [.caseInsensitive, .dotMatchesLineSeparators]
        )
        guard let match = pattern.firstMatch(in: html, range: NSRange(html.startIndex..., in: html)),
              let range = Range(match.range(at: 1), in: html) else { return nil }
        let value = String(html[range]).trimmingCharacters(in: .whitespacesAndNewlines)
        return value.isEmpty ? nil : value
    }

    private static func extractMetaContent(from html: String, name: String) -> String? {
        let pattern = try! NSRegularExpression(
            pattern: #"<meta\s+name\s*=\s*"\#(name)"\s+content\s*=\s*"([^"]*)""#,
            options: .caseInsensitive
        )
        guard let match = pattern.firstMatch(in: html, range: NSRange(html.startIndex..., in: html)),
              let range = Range(match.range(at: 1), in: html) else { return nil }
        let value = String(html[range]).trimmingCharacters(in: .whitespacesAndNewlines)
        return value.isEmpty ? nil : value
    }
}

File: app/lib/core/formats/html/html_parser.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:path/path.dart' as p;

import '../../models/reader_content_profile.dart';
import '../../services/reader_content_resolver.dart';
import '../../services/book_service.dart';

/// Metadata extracted from an HTML file.
class HtmlMetadata {
  final String title;
  final String author;
  final int chapterCount;
  final String? languageTag;
  final ContentTextDirection? textDirection;

  const HtmlMetadata({
    required this.title,
    required this.author,
    required this.chapterCount,
    this.languageTag,
    this.textDirection,
  });
}

class _HtmlChapter {
  final String title;
  final String content;

  const _HtmlChapter({required this.title, required this.content});
}

/// Parser for HTML book files (.html, .htm, .xhtml).
///
/// Extracts title from <title> tag, splits by <h1>/<h2> headings into chapters.
/// Content is served as-is to the WebView reader (already HTML).
class HtmlParser {
  static final _cache = <String, List<_HtmlChapter>>{};

  /// Extract metadata from an HTML file.
  static HtmlMetadata extractMetadata(Uint8List bytes, String filePath) {
    final html = _decode(bytes);
    final chapters = _parseChapters(html, filePath);

    return HtmlMetadata(
      title: _extractTitle(html, filePath),
      author: _extractAuthor(html),
      chapterCount: chapters.length,
      languageTag: ReaderContentResolver.extractLanguageTagFromHtml(html),
      textDirection: ReaderContentResolver.directionFromDirAttribute(
        ReaderContentResolver.extractDirAttributeFromHtml(html),
      ),
    );
  }

  /// Get chapter list for reader UI.
  static Future<List<ChapterInfo>> getChapters(String filePath) async {
    final chapters = await _loadChapters(filePath);
    return chapters.asMap().entries.map((e) {
      return ChapterInfo(index: e.key, title: e.value.title);
    }).toList();
  }

  /// Get chapter content (body HTML) for WebView rendering.
  static Future<String> getChapterContent(
    String filePath,
    int chapterIndex,
  ) async {
    final chapters = await _loadChapters(filePath);
    if (chapterIndex < 0 || chapterIndex >= chapters.length) return '';
    return chapters[chapterIndex].content;
  }

  static Future<ReaderContentProfile> getContentProfile(
    String filePath,
    int chapterIndex, {
    required String? bookLanguageTag,
    required ReaderDirectionOverride directionOverride,
  }) async {
    final file = File(filePath);
    final bytes = await file.readAsBytes();
    final html = _decode(bytes);
    final chapters = _parseChapters(html, filePath);
    final safeChapterIndex = chapterIndex.clamp(
      0,
      chapters.isEmpty ? 0 : chapters.length - 1,
    );
    final chapterContent = chapters.isEmpty
        ? html
        : chapters[safeChapterIndex].content;

    return ReaderContentResolver.resolve(
      directionOverride: directionOverride,
      explicitLanguageTag: ReaderContentResolver.extractLanguageTagFromHtml(
        html,
      ),
      explicitDir: ReaderContentResolver.extractDirAttributeFromHtml(html),
      bookLanguageTag: bookLanguageTag,
      textSample: chapterContent,
    );
  }

  static void clearCache(String filePath) => _cache.remove(filePath);

  // ── Private helpers ──────────────────────────────────────────────────────

  static Future<List<_HtmlChapter>> _loadChapters(String filePath) async {
    if (_cache.containsKey(filePath)) return _cache[filePath]!;

    final file = File(filePath);
    final bytes = await file.readAsBytes();
    final html = _decode(bytes);
    final chapters = _parseChapters(html, filePath);
    _cache[filePath] = chapters;
    return chapters;
  }

  static String _decode(Uint8List bytes) {
    try {
      return utf8.decode(bytes);
    } catch (_) {
      return latin1.decode(bytes);
    }
  }

  static String _extractTitle(String html, String filePath) {
    final match = RegExp(
      r'<title[^>]*>(.*?)</title>',
      caseSensitive: false,
      dotAll: true,
    ).firstMatch(html);
    if (match != null) {
      final title = match.group(1)?.trim() ?? '';
      if (title.isNotEmpty) return _stripTags(title);
    }
    return p.basenameWithoutExtension(filePath);
  }

  static String _extractAuthor(String html) {
    // Try meta tag: <meta name="author" content="...">
    final match = RegExp(
      r'<meta\s+name\s*=\s*"author"\s+content\s*=\s*"([^"]*)"',
      caseSensitive: false,
    ).firstMatch(html);
    return match?.group(1)?.trim() ?? '';
  }

  /// Extract body content from full HTML document.
  static String _extractBody(String html) {
    final bodyMatch = RegExp(
      r'<body[^>]*>(.*)</body>',
      caseSensitive: false,
      dotAll: true,
    ).firstMatch(html);
    return bodyMatch?.group(1)?.trim() ?? html;
  }

  /// Split HTML into chapters by <h1> or <h2> headings.
  static List<_HtmlChapter> _parseChapters(String html, String filePath) {
    final body = _extractBody(html);

    // Find all h1/h2 tags as chapter boundaries
    final headingPattern = RegExp(
      r'<(h[12])[^>]*>(.*?)</\1>',
      caseSensitive: false,
      dotAll: true,
    );

    final matches = headingPattern.allMatches(body).toList();

    if (matches.length < 2) {
      // Single chapter: the whole document
      return [
        _HtmlChapter(title: _extractTitle(html, filePath), content: body),
      ];
    }

    final chapters = <_HtmlChapter>[];

    // Content before first heading
    if (matches.first.start > 0) {
      final preContent = body.substring(0, matches.first.start).trim();
      if (preContent.isNotEmpty) {
        chapters.add(_HtmlChapter(title: '', content: preContent));
      }
    }

    for (int i = 0; i < matches.length; i++) {
      final start = matches[i].start;
      final end = i + 1 < matches.length ? matches[i + 1].start : body.length;
      final content = body.substring(start, end).trim();
      final title = _stripTags(matches[i].group(2) ?? '');

      if (content.isNotEmpty) {
        chapters.add(_HtmlChapter(title: title, content: content));
      }
    }

    return chapters.isEmpty
        ? [_HtmlChapter(title: _extractTitle(html, filePath), content: body)]
        : chapters;
  }

  static String _stripTags(String html) {
    return html.replaceAll(RegExp(r'<[^>]*>'), '').trim();
  }
}
